### DISCUSSION TITLE
Cyclic mapping with several depending DTOs

### DISCUSSION THREAD

**Discussion by Kellrkind** [mapstruct/mapstruct#2814] [Q&A] [answered]
Hey everyone,

I'm totally new to mapstruct and hope you guys can help me :)
I am building an application and have a problem with bidirectional relations in mapstruct.

Let's start with my nested entities (I'm just listing the entities with their relations between them):
- Hexe
```
@OneToMany(mappedBy = "hexe")
private List<Scan> scanList;

@OneToMany(mappedBy = "hexe")
private List<Bestellung> bestellungList;
```

- Umzug
```
@OneToMany(mappedBy = "umzug")
private List<Scan> scanList;
```

- Scan
```
@ManyToOne
@JoinColumn(name = "hexe_id", nullable = false)
@NotNull
private Hexe hexe;

@ManyToOne
@JoinColumn(name = "umzug_id", nullable = false)
@NotNull
private Umzug umzug;
```

- Bestellung
```
@ManyToOne
@JoinColumn(name = "rechnung_id", nullable = false)
@NotNull
private Rechnung rechnung;

@ManyToOne
@JoinColumn(name = "hexe_id", nullable = false)
@NotNull
private Hexe hexe;

@ManyToOne
@JoinColumn(name = "produkt_id", nullable = false)
@NotNull
private Produkt produkt;
```

- Produkt
```
@OneToMany(mappedBy = "produkt")
private List<Bestellung> bestellungList;
```

- Rechnung
```
@OneToMany(mappedBy` = "rechnung")
private List<Bestellung> bestellungList;
```

Of course I use DTO's ;)

Each of these entities has at least one relation to one or more others. In sum, you can go around in circles forever when mapping and this produces a stackoverflow-error in the end. Therefore I have pro forma deactivated all child relations in the mappers. For example in the mapper class "HexeMapper":

```
@Mapping(target = "scanList", ignore = true)
@Mapping(target = "bestellungList", ignore = true)
```

Works fine for me but I don't want to make a service call every time I want to fetch some child relations...

Now I am thinking about how to manage this, if I want to have child relations and if these should contain certain child relations.
My thought was to solve this via "@mapping". For example in the "Hexe-Mapper" it could look like this:

```
@Mapping(target = "scanList.hexe", ignore = true)
@Mapping(target = "scanList.umzug", ignore = true)
@Mapping(target = "bestellungList", ignore = true)
```

The actual goal is to define (for the first time) the following strategies:
```
@Mapping(target = "scanList.umzug.scanList", ignore = true)
@Mapping(target = "scanList.hexe", ignore = true)
@Mapping(target = "bestellungList.hexe", ignore = true)
@Mapping(target = "bestellungList.produkt.bestellungList", ignore = true)
@Mapping(target = "bestellungList.rechnung.bestellungList", ignore = true)
```
   
But this does not work so unfortunately. The "target" is now completely ignored.
I did find a way via "qualifiedBy", but that would cost me incredibly many methods, thus time and make everything confusing....

How can I achieve the desired goal without having to create thousands of individual methods?

Many thanks and have a great weekend
Martin

**Comment by Zegveld** [comment, marked as answer]
You can use 2 before mappings to handle this. For an example see [mapstruct-examples example with cycles](https://github.com/mapstruct/mapstruct-examples/tree/master/mapstruct-mapping-with-cycles/src/main/java/org/mapstruct/example/mapper).

`CycleAvoidingMappingContext` contains 2 before mappings:
* The first before mapping returns the mapped object if it has already been mapped before.
* The second before mapping stores the newly created target object for this source.

in the `EmployeeMapper` you can see an example of it being applied.

**Reply by Kellrkind** [reply]
But can I also define the depth in this way?

Say like in my example:
`@Mapping(target = "bestellungList.produkt.bestellungList", ignore = true)`

Here I go from the top mapping level 2 entities deeper I want to load and then it's over.
Hexe -> Bestellung (as List) -> Produkt -> finished

**Reply by Zegveld** [reply]
You cannot define the depth, if mapstruct arrives at the same object it will instead use the one it previously stored.
Storing takes place before the actual mapping process, this way even incompletely mapped objects can be found and referenced.

So in your example:
1. (initial) Map `bestellungList`
2. (bestellungList) Before check, is this already mapped: result false.
3. (bestellungList) create the `targetBestellungList`
4. (bestellungList) Before action, store the `targetBestellungList` as the result for this `bestellungList`.
5. (bestellungList) Map `produkt`
6. (bestellungList.produkt) Before check, is this already mapped: result false.
7. (bestellungList.produkt) create the `targetProdukt`
8. (bestellungList.produkt) Before action, store the `targetProdukt` as the result for this `produkt`.
9. (bestellungList.produkt) Map `bestellungList`
10. (bestellungList.produkt.bestellungList) Before check, is this already mapped: result true
11. (bestellungList.produkt.bestellungList) return stored `targetBestellungList`
12. (bestellungList.produkt) do other `targetProdukt` mappings
13. (bestellungList.produkt) return `targetProdukt`
14. (bestellungList) do other `targetBestellungList` mappings
15. (bestellungList) return `targetBestellungList`
16. (initial) receive map result: `targetBestellungList`

This result `targetBestellungList` contains a `targetProdukt` that points back to `targetBestellungList`.
Cycle is present, but mapstruct didn't end up in an infinite loop.

**Reply by Kellrkind** [reply]
First of all thank you for your time and your help :) 💪🏻 

Sounds logically but I can't imagine yet how to implement this 🙈 
I'm going to give it a try the next days but I'm not very optimistic 😅

**Reply by Zegveld** [reply]
you can most likely use the `CycleAvoidingMappingContext` from [the example repository](https://github.com/mapstruct/mapstruct-examples/tree/master/mapstruct-mapping-with-cycles/src/main/java/org/mapstruct/example/mapper) without changing it to get it working. Add it as an additional argument in your mapping methods and mark it with `@Context` that should basically do it.

Good luck.

**Reply by Kellrkind** [reply]
I've understood that part with the context integration.
What's not clear for me is how to use it in my situation 🙈 

Should I replace the <T> with my own entity or leave it as it is?

And do i have to do call the other mappers manually and implement therefore an additional context in the specific mapper and so on or how do I handle it?

**Reply by Zegveld** [reply]
all manually defined mapping methods in the mappers would gain the `@Context CycleAvoidingMappingContext context` parameter.
and all calls to these mapping methods would gain `new CycleAvoidingMappingContext()` to fill in that parameter, unless it already receives the `CycleAvoidingMappingContext` object then it just passes that one on.

the `CycleAvoidingMappingContext` class itself from the example repository will automatically work for all mappers, no need to alter that one.

**Reply by Kellrkind** [reply]
Unfortunately my code won't compile after implementing the CycleAvoidingMappingContext. 

```
[ERROR] /Users/martin.keller/Desktop/workspace/narrenzunft/server/target/generated-sources/annotations/com/kellrkind/narrenzunft_server/mapper/HexeMapperImpl.java:[33,48] incompatible types: inference variable T has incompatible bounds
[ERROR]     equality constraints: com.kellrkind.narrenzunft_server.models.Hexe.HexeBuilder
[ERROR]     lower bounds: com.kellrkind.narrenzunft_server.models.Hexe,java.lang.Object
```

The generated HexeMapperImpl always says `incompatible types`. Same for every other mapper in which I've implemented the CycleAvoidingMappingContext...

Do you have any idea why the builder is used in the generated mapper instead of the object itself?

![image](https://user-images.githubusercontent.com/7069944/163011062-c39a980f-89e1-4141-8e17-e969c0ff5b0d.png)

**Reply by Zegveld** [reply]
if mapstruct detects a builder it tries to use that to create the object. In your case this is not wanted so you can disable the builder detection
for example: `@Mapper(builder = @Builder(disableBuilder = true))`

**Reply by Kellrkind** [reply]
It fucking works! Holy moly!

Thank you so much for the effort you put in. 

I'd be happy to reciprocate.
Do you have a dogecoin wallet? If so, I'd be happy to send you some Doge as a thank you :)

**Comment by Kellrkind** [comment]
For completion. The suggestion from @Zegveld works perfectly fine for me :)

Here is my HexeMapper:
```
import com.kellrkind.narrenzunft_server.mapper.context.CycleAvoidingMappingContext;
import com.kellrkind.narrenzunft_server.models.Hexe;
import com.kellrkind.narrenzunft_server.models.dto.HexeDTO;
import org.mapstruct.Builder;
import org.mapstruct.Context;
import org.mapstruct.InheritInverseConfiguration;
import org.mapstruct.Mapper;

@Mapper(builder = @Builder(disableBuilder = true))
public interface HexeMapper {

   Hexe dtoToPojo(HexeDTO destination, @Context CycleAvoidingMappingContext context);

   @InheritInverseConfiguration
   HexeDTO pojoToDTO(Hexe source, @Context CycleAvoidingMappingContext context);

}
```

Here is the CycleAvoidingMappingContext (same as the on mentioned in the example):
```
import java.util.IdentityHashMap;
import java.util.Map;
import org.mapstruct.BeforeMapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.TargetType;

public class CycleAvoidingMappingContext {

   private final Map<Object, Object> knownInstances = new IdentityHashMap<>();

   @BeforeMapping
   public <T> T getMappedInstance(Object source, @TargetType Class<T> targetType) {
      return (T) knownInstances.get(source);
   }

   @BeforeMapping
   public void storeMappedInstance(Object source, @MappingTarget Object target) {
      knownInstances.put(source, target);
   }
}
```

And in the specific places I just do something like this:
`hexeDTOList.add(mapper.pojoToDTO(hexe, new CycleAvoidingMappingContext()));`

Same same in every other mapper which is included in my cyclic mapping
